federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	service.Name = meta.Name
+	service.GenerateName = fedService.Name
+	service.ResourceVersion = ""
+
+	api.GenerateName(api.SimpleNameGenerator, meta)
+	// TODO: to generate labels based on hash
+
+	// handle available clusters one by one
+	for clusterName, client := range s.clusterClientMap {
+		go s.processServiceForCluster(key, clusterName, service, appliedState, client)
+	}
+}
+
+func retrieveExternalIPs(client clientset.Clientset, service v1.Service) {
+	// TODO: sleep and pull service info to check LB changes? or create k8s cluster service watch when create clusterClientMap
+	for i := 0; i < clientRetryCount; i++ {
+		time.Sleep(clientRetryInterval)
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			return svc.Spec.ExternalIPs
+		}
+	}
+}
+
+func getServiceByLabel(client clientset.Clientset, service v1.Service) v1.Service{
+	// TODO: query service by service-spec-hash
+	svc, err := client.Core().Services(service.Namespace).List(&unversioned.LabelSelector{MatchLabels: map[string]string{serviceHashLabelName: "baz"}})
+	// One service in one cluster with same hash
+	if err == nil && svc != nil && len(svc.Items) == 0 {
+		return svc.Items[0]
+	} else {
+		return nil
+	}
+}
+
+func (s *ServiceController) ensureClusterServices(service *v1.Service, client clientset.Clientset, clusterName string) ([]string, error) {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			// service exists
+			glog.V(4).Infof("service exists in cluster %v, trying to update", clusterName)
+			_, err = client.Core().Services(service.Namespace).UpdateStatus(service)
+			if err == nil {
+				// Update successfully, get LB IPs
+				return retrieveExternalIPs(client, service)
+			}
+		}
+
+		if errors.IsNotFound(err) {
+			// Create service if is not found
+			glog.Infof("Service '%s/%s' does not exist in %s, trying to create new",
+				service.Namespace, service.Name, clusterName)
+			_, err = client.Core().Services(service.Namespace).Create(service)
+			if err == nil {
+				return retrieveExternalIPs(client, service)
+			}
+		}
+
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
+				service.Namespace, service.Name, err)
+			// ignore and get LB info
+			return retrieveExternalIPs(client, service)
+		}
+
+		glog.Warningf("Failed to get LoadBalancerStatus for service '%s/%s' after creating it in cluster %s: %v",
+			service.Namespace, service.Name, clusterName, err )
+		time.Sleep(clientRetryInterval)
+	}
+	return err
+}
+
+func (s *ServiceController) ensureDNSRecords(clusterName string, service *v1.Service) error {
+	for _, zone := range s.zones {
+		ResourceRecordSets, ok := zone.ResourceRecordSets()
+		if !ok {
+			//zone name or cluster name?
+			rrsList, err := ResourceRecordSets.List()
+			if err != nil {
+				return fmt.Errorf("fail to get rrs")
```

Comment:
Please don't suppress errors like this.  Return the original error (err), or as a minimum include the original error in your error message.